### PR TITLE
zephyr-ipc: Invoke platform_pm_runtime_power_off during D3

### PR DESCRIPTION
**Pull request by ranj063**
When entering D3, call the platform_pm_runtime_power_off to prevent the IPC timeouts seen with the CTX_SAVE IPC seen on the host.

Signed-off-by: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>

### CODE REVIEW

File: src/ipc/ipc-zephyr.c
Comment (kv2019i) [inline src/ipc/ipc-zephyr.c:129]: @tmleman @marcinszkudlinski Any thoughts how this should be handled? This feels a bit wrong to call to platform_pm_runtime_power_off() with the native driver, but somehow we need to handle the IPC response in this case (e.g. the code in src/platform/intel/cavs/lib/power_down.S for cAVS platforms -- send IPC reply and shut down).

**Review by kv2019i — COMMENTED** [review]
Let's check that we are aligned with eventual target design here, pinged @tmleman and @marcinszkudlinski

File: src/ipc/ipc-zephyr.c
Comment (lgirdwood) [reply src/ipc/ipc-zephyr.c:129]: I would expect here that we would have to call the Zephyr native API for shutdown() which should permit return and then allow the assembly IPC logic to send the IPC reply ? adding @nashif

**Review by lgirdwood — COMMENTED** [review]
I'm good for this in the current state if needed for unblocking driver. If we need to revisit (e.g. for native APIs and any otehr changes) then we can add a TODO: comment and a feature to track in GH.

File: src/ipc/ipc-zephyr.c
Comment (kv2019i) [reply src/ipc/ipc-zephyr.c:129]: Aa, Ranjani found the logic is coming in this PR -> https://github.com/thesofproject/sof/pull/6440

Comment (ranj063) [reply src/ipc/ipc-zephyr.c:129]: but we cannot use the same logic for TGL right @kv2019i. We need an zephyr native API for shutdown like @lgirdwood suggested. But until then my patch can be stop gap

**Comment by aborisovich** [comment]
I've tried this solution for MTL and it did not work.
During execution of the IPC task in another thread, other IPC message can be handled in the interrupt context synchronously.
Please note that body of `ipc_platform_do_cmd` and `ipc_cmd` do not use the spinlock from `z_intel_adsp_ipc_isr` function from `zephyr\soc\xtensa\intel_adsp\common\ipc.c` , what may lead to a race in setting IPC message completion acknowledgement bits - clearing IDA BUSY bit what indicates that Sideband peer agent has completed the operation. You get away with it for TGL because power_down.S assembly differs for TGL and MTL. Just compare body of `_PD_SEND_IPC` from power_down.S:
image

Overall, this can be merged as a temporary work-around if this fixes TGL but I would double check whether solution from my PR #6440 cannot be used for fixing TGL. After all TGL with native drivers overlay has Zephyr power manager enabled.
Please double check whether this work-around is really mandatory. If not I will remove [DO NOT MERGE] from my PR and merge it as fix for TGL adding MTL specific code later.

**Comment by kv2019i** [comment]
Thanks @aborisovich for the analysis. I think we are safe as upper layer protocol ensures there will not be a new IPC message from the host, until previous one is handled. And for notifications (from DSP to host), all the async one are send as "not high priority", i.e. they go through the worker thread and it has:

»       if (ipc_get()->pm_prepare_D3)
»       »       goto out;

... so postponing sending if system is going to D3.

File: src/ipc/ipc-zephyr.c
Comment (kv2019i) [reply src/ipc/ipc-zephyr.c:129]: Ack, the right solution is of course to implement PM support for TGL as well, but in spirit of our incremental move from XTOS drivers to native drivers, having a intermediate solution here for TGL is fine.

**Review by kv2019i — APPROVED** [review]
If this doesn't "uglify" the native solution with Zephyr PM (now down for MTL), and @aborisovich seems to indicate it doesn't, I'm good to merge this.

**Review by aborisovich — CHANGES_REQUESTED** [review]
This can be merged as solution for TGL but MTL really should not have `platform_pm_runtime_power_off` function implementations. Please remove those dummy function calls from `ace/lib/pm_runtime.c` and `pm_runtime.h` for meteorlake and wrap `platform_pm_runtime_power_off()` invocation with `#ifdef CAVS` family.

**Review by tmleman — APPROVED** [review]

**Comment by ranj063** [comment]
> This can be merged as solution for TGL but MTL really should not have `platform_pm_runtime_power_off` function implementations. Please remove those dummy function calls from `ace/lib/pm_runtime.c` and `pm_runtime.h` for meteorlake and wrap `platform_pm_runtime_power_off()` invocation with `#ifdef CAVS` family.

done @aborisovich

**Review by aborisovich — APPROVED** [review]

**Pull request merged**
